### DISCUSSION TITLE
Could there be an Equalizer?

### DISCUSSION THREAD

**Discussion by TreeTrunk221** [UnicornsOnLSD/finamp#717] [General]
Could you add an equalizer? For reference, the app Musi has one that is pretty good.

**Comment by othyn** [comment]
Same. I just went hunting for one on iOS and it appears there currently isn't a way to do this?

My headphones sound a little bright and I would like to tweak a few things.

**Comment by Chaphasilor** [comment]
Like with so many things the answer is: only on Android  

We use a plugin called [`just_audio`](https://pub.dev/packages/just_audio) for audio playback, which handles most of the different platforms (Android, iOS, etc.) for us, in a relatively consistent way. This vastly simplifies things for us, otherwise such a cross-platform audio player would be even more work.

Sadly, `just_audio` only supports equalizers on Android, not iOS or other platforms. I actually think there's no way to configure an equalizer into the media player on iOS at all, aside from building everything yourself. This only works for apps that directly deal with audio samples, like Logic Pro, because they have a different audio pipeline than simple media players.

On Android there's also a system-wide equalizer that can be used instead, so there's less need for the feature. And adding this feature just for Android currently isn't worth the effort in my opinion.

That being said, if anyone wants to give this a try, I'd be happy to accept a PR!  
The code would be very similar to what is being used for the volume normalization, so the following section should contain almost all needed "concepts". Building the UI for configuring the equalizer will be the most challenging part though.

https://github.com/jmshrv/finamp/blob/5a33802d4f0d782c1d588c24faf8aa0c2e9e66e1/lib/services/music_player_background_task.dart#L53-L121

**Reply by othyn** [reply]
Thank you for the detailed answer, very interesting!

I know PlexAmp is closed source, but any ideas on how they achieve their cross platform EQ? Is it an in-house home grown implementation? I suppose to know that we'd need to know their audio pipeline, and that circles back to the problem of it being closed source.

It's highly annoying that iOS doesn't have a system level EQ. I went hunting for it before remembering that the one in Settings is only for the in built Music app... thanks Apple. Given their bread and butter is music production hardware/OS, its a strange omission. But then again, thats patently Apple at this point. A system level EQ would solve this for sure.

There was an original PR opened for this, then a subsequent one which overtook it:

1. https://github.com/ryanheise/just_audio/pull/658
2. https://github.com/ryanheise/just_audio/pull/784

But it appears the PR has been left hanging.

For reference, this is the one in PlexAmp:

![image](https://github.com/jmshrv/finamp/assets/7256684/232f987b-bfb3-4d09-a58b-fe85a4bae6ca)

**Reply by Chaphasilor** [reply]
Yes, I'm guessing they are either using a custom implementation, or a more advanced audio pipeline, because Apple has kept that feature for their own apps. If that just_audio PR ever gets merged, this is definitely a feature I'd be willing to implement!

**Reply by langbl28** [reply]
Came to ask for the feature but stayed for the great explanation of what’s possible. Thank you for your work. Hope to see an EQ someday if it’s ever more achievable!

**Reply by UnicornsOnLSD** [reply]
Having a poke about with the iOS version of Plexamp, both [AudioToolbox](https://developer.apple.com/documentation/audiotoolbox) and [AVKit](https://developer.apple.com/documentation/avkit), whereas `just_audio` uses AVFoundation.

![Plexamp importing AVKit and AudioToolbox](https://github.com/user-attachments/assets/660bcdc2-7dcf-4284-84cc-d6e70fcd73d1)

AudioToolbox is a much more low-level API, so it's likely that Plexamp is doing some very custom audio handling, while `just_audio` is just using the built in player from iOS/macOS.

With a bit more digging (searching the program text for OPUS to see if libopus was anywhere lol), I found references to [BASS](https://www.un4seen.com/), so it's likely that Plexamp is using this library to feed into iOS via AudioToolbox. I'm no reverse engineer so I can't really verify it (also Plexamp is a React Native app so the code is a bit all over the place), but this is probably enough speculation for now :)

![Search results for OPUS, showing BASS and many other log messages related to OPUS streams](https://github.com/user-attachments/assets/cae073b8-5980-4cd7-afe4-dd562f4122f2)

Fun fact: one of the reasons I started making Finamp was because Plexamp's Android app couldn't handle surround FLACs (it'd just drop channels that weren't L and R), so the few SACDs I had in my collection were unlistenable 🙃

**Reply by UnicornsOnLSD** [reply]
Forgot to say - with this, Plexamp has full control over audio samples, so can do EQ on its own

**Reply by UnicornsOnLSD** [reply]
and yep, looks like BASS handles EQ on its own

![Ghidra screenshot showing BASS EQ functions](https://github.com/user-attachments/assets/94fdfb7f-2d4d-4c26-b662-b04c70691f13)

**Reply by UnicornsOnLSD** [reply]
(^ @othyn, just in case you're not watching this thread)

**Reply by othyn** [reply]
That's a super interesting deep dive, thank you for sharing! I really should brush up on some reverse engineering, as it's really awesome to see what you've been able to find.
